### PR TITLE
[WIP] js: tighten up internalize and externalize

### PR DESCRIPTION
**Pull request by myitcv**
**!! DO NOT REVIEW !!**

**NEEDS REBASE AFTER #687 IS MERGED**

### Status

Step 1 and 2 (see below) review.

### Description

A work-in-progress PR to complement the discussion in #617 

Tighten up `$internalize` and `$externalize` such that:

* zero val translation of `undefined` and `null` is well defined
* the domain of valid values for a given target type is well defined
* runtime errors are thrown for invalid values

See impact of this PR on `honnef.co/go/js/dom` in [this comparison](https://github.com/dominikh/go-js-dom/compare/master...myitcv:internalize_zero_val?expand=1)

### Documentation

Proposed human-readable definitions of `$internalize` and `$externalize` can be seen [here](https://github.com/myitcv/gopherjs-compiler/blob/internalize_zero_val/js/js.go).

### Scope of `$internalize` testing

The domain of source values we need to consider is:

* `null`
* `undefined` (there is no need to distinguish between a field/variable being explicitly assigned the value of `undefined` and it being so by virtue of not having been initialised)
* Primitive and non-primitive values (see [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String#Distinction_between_string_primitives_and_String_objects))

### Overall PR TODO list

1. Agree on structure of code, tests and docs for a couple of types, e.g. `string` and `bool`
2. Work out and fix reflection (manifests via `TestReflection` failing in `js` tests)
3. Complete revised implementation of `$internalize` and `$externalize` for all types to allow for testing that code in real-life code bases (not just tests)
4. Finish docs for all types
5. Tighten up wording of `*js.Object`-special struct types in `js` package (unless we work this into the tables for `$internalize` and `$externalize`
6. Document ways in which you can fallback to Javascript "loose" type conversions

### Consideration for later PRs

* Splitting `$internalize` and `$externalize` into separate, smaller type-specific functions 
* Fixing the fact that `null` and `undefined` do not internalize as `(*T)(nil)` for a `*js.Object`-special struct type `T`

### CODE REVIEW

File: js/js_test.go [hunk 1]
```diff
@@ -571,3 +571,20 @@ func TestUint8Array(t *testing.T) {
 		t.Errorf("Non-empty byte array is not externalized as a Uint8Array")
 	}
 }
+
+func TestInternalizeZeroVal(t *testing.T) {
+	type S struct {
+		*js.Object
+
+		string `js:"string"`
```
Comment (shurcooL) [inline js/js_test.go]: Since this is a test of internalization logic (as I understand from our conversation in #617), I would suggest rewriting it to avoid using `js` struct field tags. I think it's simpler to use `Object.String` method, it's lower level and more explicit, making it more clear what the test is testing.

Comment (myitcv) [inline js/js_test.go]: @shurcooL I think for completeness we'll need to test `String()` as well, but in order that we can cover all Go types I think we need to follow the struct field tags approach... unless I'm missing something?

Comment (shurcooL) [inline js/js_test.go]: > but in order that we can cover all Go types I think we need to follow the struct field tags approach...

Are methods like `String`, `Float`, `Bool` `Uint64`, etc., not sufficient? What's not possible without struct field tags?

Comment (myitcv) [inline js/js_test.go]: All the types that do not have corresponding methods on `Object`. e.g. the specific numeric types like `int8`.

Comment (shurcooL) [inline js/js_test.go]: Can `Interface().(int8)` be done for those, or is that not equivalent?

Comment (myitcv) [inline js/js_test.go]: I think `Interface()` is another test case altogether because it [uses the Javascript type of the value to decide which Go type to resolve to](https://github.com/gopherjs/gopherjs/blob/3496c6f94e1b945f3fc9580f6982675c0a74cd6c/compiler/prelude/jsmapping.go#L258-L308). So in the context of this discussion, I suspect use of `Interface()` wouldn't actually be a problem per se.

What you've proposed transpiles to:

```javascript
$assertType($internalize(o.A, $emptyInterface), $Int8)
```

So we still need to have test cases for all the Go types... because it's `$internalize` that we're ultimately testing here.

Comment (shurcooL) [inline js/js_test.go]: Got it, thanks. Then it's okay to use the struct field tags, but my request is to document that what's being tested is internalize behavior, not the struct field tag behavior.

Comment (myitcv) [inline js/js_test.go]: Ah I see; 100% agree on that point.

File: js/js.go [hunk 2]
```diff
@@ -1,28 +1,47 @@
...
+//
+// Source values annotated with † are generally applicable to all valid
+// values of the target type. e.g. for target type string, "ok" represents
+// all valid string values.
+//
+// Externalization
+//
+// When values pass from Go to Javascript, a process known as externalization,
+// the following conversion table is applied:
+//
```
Comment (flimzy) [inline js/js.go]: Table missing?

File: js/js.go [hunk 3]
```diff
@@ -1,28 +1,47 @@
...
+// target type if a conversion is attempted, e.g. a Javascript number value
+// being assigned to a string type Go variable.
+//
+// Source values annotated with † are generally applicable to all valid
+// values of the target type. e.g. for target type string, "ok" represents
+// all valid string values.
+//
+// Externalization
+//
+// When values pass from Go to Javascript, a process known as externalization,
```
Comment (flimzy) [inline js/js.go]: Incomplete sentence

Comment (myitcv) [inline js/js.go]: Thanks @flimzy - I should have mentioned this entire PR is still very much WIP.

In particular, I haven't even started on the `$externalize` side of things. Had some fairly interesting edge cases on the `$internalize` side of things

File: js/js.go (hunk 2 above)
Comment (shurcooL) [inline js/js.go]: See next line.

**Comment by myitcv** [comment]
@flimzy @shurcooL - thanks both for taking a look at this.

I should have been clearer in my update message in #617 because I value your time... This is still WIP so I will give you a shout when it's essentially "ready" for review. There are still too many rough edges. 

My plan is to get a couple of types (probably `string` and `bool`) done for both `$internalize` and `$externalize` and then get some feedback on how things look, from code and tests through to docs etc.

Just wanted to post an update to confirm I'm still "on" this, albeit after a significant period of inactivity!

**Comment by shurcooL** [comment]
> I should have been clearer in my update message in #617 because I value your time... This is still WIP so I will give you a shout when it's essentially "ready" for review. There are still too many rough edges.

No problem.

I suggest using Go code review convention for achieving this, put a "DO NOT REVIEW" text anywhere in the PR description to signal that. (See https://groups.google.com/d/msg/golang-dev/YU56eKYiXJg/K1CkXo7Iy0gJ.)

**Comment by myitcv** [comment]
> I suggest...

Done. I've switched from using my `[WIP]` convention in the title to your suggestion in the PR description.

Thanks again

**Comment by myitcv** [comment]
@shurcooL @flimzy et al.

Given there's quite a lot to work through in this PR, I've updated the description in this PR with, amongst other things, the following:

* Status
* Overall PR TODO list

Would welcome your thoughts on the TODO list, in terms of content but also approach (i.e. using the TODO list as a means of communicating how I plan to go about things).

At this point I think I've got to the end of step 1, hence am marking this for a review.

All thoughts/feedback welcomed.

Thanks

**Comment by myitcv** [comment]
Worth pointing out that at this stage, not least because of point 2 from the TODO list, I'm not expecting the build to pass.

File: js/js.go [hunk 4]
```diff
@@ -1,28 +1,60 @@
...
+//  |                |               | undefined               | ""     |
+//  |                |               | ""                      | ""     |
+//  |                |               | new String("")          | ""     |
+//  |                |               | "ok" †                  | "ok"   |
+//  |                |               | new String("ok") †      | "ok"   |
+//  |----------------+---------------+-------------------------+--------|
+//  | bool           | none          | null                    | false  |
+//  |                |               | undefined               | false  |
+//  |                |               | false †                 | false  |
+//  |                |               | new Boolean(false) †    | false  |
```
Comment (flimzy) [inline js/js.go]: Is it worth including `true` examples here, for the sake of completeness, similar to the non-zero value examples for strings?

File: js/js_test.go [hunk 5]
```diff
@@ -467,7 +467,7 @@ func TestCallWithNull(t *testing.T) {
 func TestReflection(t *testing.T) {
 	o := js.Global.Call("eval", "({ answer: 42 })")
 	if reflect.ValueOf(o).Interface().(*js.Object) != o {
-		t.Fail()
+		t.Fatal()
```
Comment (flimzy) [inline js/js_test.go]: AFAIK, `Fatal()` is for reporting a failure message.  Maybe `FailNow()` would be more semantically correct?

File: js/js_test.go [hunk 6]
```diff
@@ -571,3 +572,207 @@ func TestUint8Array(t *testing.T) {
...
+
+	string string `js:"string"`
+	bool   bool   `js:"bool"`
+}
+
+func TestInternalizeString(t *testing.T) {
+	fieldName := "string"
+	zero := ""
+
+	s := &Internalize{Object: js.Global.Get("Object").New()}
```
Comment (flimzy) [inline js/js_test.go]: Stylistic note:  I wonder if a table-based test would make sense here?  It may make the tests more readable, reduce the need for the `// ***` comment blocks, and if coupled with `t.Run(...)` calls for each test, could make the output more readable.

File: js/js_test.go (hunk 5 above)
Comment (shurcooL) [inline js/js_test.go]: Is it not possible to include a helpful error message in the `t.Fatal` call? If possible, that's better.

File: js/js_test.go (hunk 6 above)
Comment (myitcv) [inline js/js_test.go]: You raise a good point. I half considered this myself, but postponed giving it much thought until we've confirmed we're actually testing the right stuff... at which point extracting to table tests should be relatively easy. 

The little thought I did give it however amount to the following: I'm not convinced we _can_ actually use table tests. We need to address specific fields of the `Internalize` struct and do so in such a way that the `gopherjs` compiler sees us doing so and compiles to the corresponding Javascript. Put another way, the code we write in these tests needs to correspond _exactly_ to the code that will be written in the wild.

Therefore the one approach I have considered fairly seriously is code generating the tests. Because they do follow a formula that's for sure.

But as I said, I haven't given this much time/though so would very much welcome any suggestions.

File: js/js_test.go (hunk 5 above)
Comment (myitcv) [inline js/js_test.go]: To be honest I was being a little lazy here in just flipping the existing `t.Fail()` to `t.Fatal()` to help with debugging step 2 from the TODO. 

I'm sure we could add helpful messages to the `t.Fatal()` calls... because the existing `t.Fail()` calls are equally lossy. Could even be something we do in another PR.

File: js/js.go (hunk 4 above)
Comment (myitcv) [inline js/js.go]: We would end up adding two lines: `true` and `new Boolean(true)` - hence why I opted for the † callout.

File: js/js_test.go (hunk 5 above)
Comment (shurcooL) [inline js/js_test.go]: It's usually easier to do these minor things as you're working in a relevant context than later. But this is completely up to you, and not critical for getting the PR into a mergeable state.

File: js/js_test.go (hunk 6 above)
Comment (myitcv) [reply js/js_test.go]: @flimzy just pushed up a code-generated version of the tests which seems to work well. Let me know what you think.

**Comment by myitcv** [comment]
Just pushed up a new commit per https://github.com/gopherjs/gopherjs/issues/617#issuecomment-324688005

**Comment by flimzy** [comment]
Nearly 4 years later, @nevkontakte and I have decided to close this PR.  We like the effort here, and some of it will be very informative if we address these issues in the future, but with limited bandwidth, and the author having left GopherJS development, we feel this PR is at a dead end.
